### PR TITLE
[PORT][8.0] stock picking mass assign

### PR DESCRIPTION
**Pull request by legalsylvain**
- port v7 last changes;
- use new api;
- change terms because module is now usefull for picking in and out;
- change module name because mass action feat more with new recent features; 

[FIX] it's now possible to mass force stock picking that are in a 'partially available' state;

Thanks for your review;

### CODE REVIEW

**Comment by lepistone** [comment]
Hi @legalsylvain,

tests fail because there is still a reference to `stock.picking.out` in a test.

**Comment by legalsylvain** [comment]
@lepistone : Thanks for your review. Fixed.

File: stock_picking_mass_action/README.rst [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    * Confirm draft picking;
+    * check availability of picking;
+    * force availability of picking;
+    * transfer picking;
+    * create invoice / refund;
+
+* A scheduled action to check availability of all the stock picking.
+  It is not active by default.
+
+This may be necessary for those who want to check the availability
```
Comment (pedrobaeza) [inline stock_picking_mass_action/README.rst]: Put this joined with the previous point, as it is related

File: stock_picking_mass_action/README.rst [hunk 2]
```diff
@@ -0,0 +1,52 @@
+Stock Picking Mass Action
+=========================
+
+Facilities to check the availability of stock picking list:
+
+* A wizard which allows on multiple stock pickings at a time to:
+    * Confirm draft picking;
```
Comment (pedrobaeza) [inline stock_picking_mass_action/README.rst:11]: Put here another line break for correct formatting of the RST

File: stock_picking_mass_action/__openerp__.py [hunk 3]
```diff
@@ -19,34 +19,21 @@
...
- "test": ['test/test_check_assign_all.yml',
-          ],
- 'installable': False,
- 'auto_install': False,
- }
+{
+    'name': 'Stock Picking Mass Action',
+    'version': '1.0',
+    'author': 'Camptocamp,GRAP,Odoo Community Association (OCA)',
+    'website': 'http://www.camptocamp.com',
```
Comment (pedrobaeza) [inline stock_picking_mass_action/__openerp__.py]: In website, point to OCA repo

File: stock_picking_mass_action/i18n/de.po [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+"Language-Team: \n"
+"MIME-Version: 1.0\n"
+"Content-Type: text/plain; charset=UTF-8\n"
+"Content-Transfer-Encoding: \n"
+"Plural-Forms: \n"
+
+#. module: stock_picking_mass_action
+#: view:stock.picking.mass.action:stock_picking_mass_action.view_stock_picking_mass_action_form
+msgid "Apply"
+msgstr "Apply"
```
Comment (pedrobaeza) [inline stock_picking_mass_action/i18n/de.po]: This isn't translated

File: stock_picking_mass_action/i18n/de.po [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+#. module: stock_picking_mass_action
+#: model:ir.actions.act_window,name:stock_picking_mass_action.action_create_invoice
+#: field:stock.picking.mass.action,create_invoice:0
+msgid "Create Invoices/Refunds"
+msgstr "Erstelle Rechnung / Gutschrift"
+
+#. module: stock_picking_mass_action
+#: field:stock.picking.mass.action,create_uid:0
+msgid "Created by"
+msgstr "Created by"
```
Comment (pedrobaeza) [inline stock_picking_mass_action/i18n/de.po]: Neither this and other terms

File: stock_picking_mass_action/models/stock_picking.py [hunk 6]
```diff
@@ -18,35 +18,32 @@
 #    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #
 ##############################################################################
+
 import logging
 
 from openerp.osv import orm
+from openerp.models import Model, api
```
Comment (pedrobaeza) [inline stock_picking_mass_action/models/stock_picking.py]: api is not inside openerp.models, but directly on openerp. Why not using usual form: `from openerp import models, api" and use`models.Model` for the class?

File: stock_picking_mass_action/models/stock_picking.py [hunk 7]
```diff
@@ -18,35 +18,32 @@
...
-            ids = [ids]
-        if ids is None:
-            domain = [('type', '=', 'out'),
-                      ('state', '=', 'confirmed')]
-            ids = self.search(cr, uid, domain,
-                              order='min_date',
-                              context=context)
+        spt_obj = self.env['stock.picking.type']
+        spt_out_lst = spt_obj.search([('code', '=', 'outgoing')])
+        spt_out_ids = [x.id for x in spt_out_lst]
```
Comment (pedrobaeza) [inline stock_picking_mass_action/models/stock_picking.py]: You don't need this. Use `spt_out_lst.ids` directly.

File: stock_picking_mass_action/models/stock_picking.py [hunk 8]
```diff
@@ -18,35 +18,32 @@
...
-        if isinstance(ids, (int, long)):
-            ids = [ids]
-        if ids is None:
-            domain = [('type', '=', 'out'),
-                      ('state', '=', 'confirmed')]
-            ids = self.search(cr, uid, domain,
-                              order='min_date',
-                              context=context)
+        spt_obj = self.env['stock.picking.type']
+        spt_out_lst = spt_obj.search([('code', '=', 'outgoing')])
```
Comment (pedrobaeza) [inline stock_picking_mass_action/models/stock_picking.py]: I prefer not cryptic variable names like out_pickings.

File: stock_picking_mass_action/i18n/de.po (hunk 4 above)
Comment (legalsylvain) [inline stock_picking_mass_action/i18n/de.po]: I don't speak deutch. It will be translated in another PR / by transiflex.

Kind regards.

File: stock_picking_mass_action/models/stock_picking.py (hunk 7 above)
Comment (legalsylvain) [inline stock_picking_mass_action/models/stock_picking.py]: thanks ! I didn't know that V8 trick.

File: stock_picking_mass_action/i18n/de.po (hunk 4 above)
Comment (pedrobaeza) [inline stock_picking_mass_action/i18n/de.po]: But this way, terms will appear in Transifex as translated.

Comment (pedrobaeza) [inline stock_picking_mass_action/i18n/de.po]: You should empty non-translated terms

File: stock_picking_mass_action/models/stock_picking.py [hunk 9]
```diff
@@ -18,35 +18,32 @@
...
+        domain = [('picking_type_id', 'in', spt_out_ids),
+                  ('state', '=', 'confirmed')]
+        records = self.search(domain, order='min_date')
 
-        for picking_id in ids:
+        for record in records:
             try:
-                self.action_assign(cr, uid, [picking_id], context)
+                record.action_assign()
             except orm.except_orm:
```
Comment (pedrobaeza) [inline stock_picking_mass_action/models/stock_picking.py]: Now you should check also `exceptions.Warning` or any of the possible exceptions. Maybe a general except?

File: stock_picking_mass_action/models/stock_picking.py [hunk 10]
```diff
@@ -18,35 +18,32 @@
...
                 # ignore the error, the picking will just stay as confirmed
-                name = self.read(cr, uid, picking_id, ['name'],
-                                 context=context)['name']
                 _logger.info('error in action_assign for picking %s',
-                             name, exc_info=True)
-        return True
```
Comment (pedrobaeza) [inline stock_picking_mass_action/models/stock_picking.py:52]: I think you should still return True

File: stock_picking_mass_action/wizard/mass_action.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+        transfert_wizard_obj = self.env['stock.transfer_details']
+        picking_ids = self.env.context.get('active_ids')
+
+        # Get draft pickings and confirm them if asked
+        if self.confirm:
+            domain = [('state', '=', 'draft'),
+                      ('id', 'in', picking_ids)]
+            draft_picking_lst = picking_obj.search(domain, order='min_date')
+            draft_picking_lst.action_confirm()
+
```
Comment (pedrobaeza) [inline stock_picking_mass_action/wizard/mass_action.py:88]: I see no need for these empty break lines

File: stock_picking_mass_action/models/stock_picking.py (hunk 9 above)
Comment (legalsylvain) [inline stock_picking_mass_action/models/stock_picking.py]: I just ported this part of code, but i don't use it. I guess the author write this line to ignore error if some product are not "assignable". we don't want to ignore all the errors.

File: stock_picking_mass_action/models/stock_picking.py (hunk 10 above)
Comment (legalsylvain) [inline stock_picking_mass_action/models/stock_picking.py:52]: Why ? return true is now unecessary in V8.

Comment (pedrobaeza) [inline stock_picking_mass_action/models/stock_picking.py:52]: Ah, OK, I wasn't sure about this part

File: stock_picking_mass_action/models/stock_picking.py (hunk 9 above)
Comment (lepistone) [inline stock_picking_mass_action/models/stock_picking.py]: I agree with @legalsylvain: we should not silence new exceptions unless we know specifically why.

Comment (pedrobaeza) [inline stock_picking_mass_action/models/stock_picking.py]: OK

File: stock_picking_mass_action/wizard/mass_action.py (hunk 11 above)
Comment (legalsylvain) [inline stock_picking_mass_action/wizard/mass_action.py:88]: tastes and colors ? or are there some OCA / python conventions about that point ?

**Comment by legalsylvain** [comment]
@pedrobaeza : Hi. Thanks for your review, I took into account most of your remarks.

Let me know if all is ok.

Kind regards.

File: stock_picking_mass_action/wizard/mass_action.py (hunk 11 above)
Comment (pedrobaeza) [inline stock_picking_mass_action/wizard/mass_action.py:88]: Only tastes, but if you're using already comments, you already have a separator, so no need for a second one

**Comment by pedrobaeza** [comment]
Thanks :+1:

File: stock_picking_mass_action/README.rst [hunk 12]
```diff
@@ -0,0 +1,53 @@
+Stock Picking Mass Action
```
Comment (yvaucher) [inline stock_picking_mass_action/README.rst:4]: Please add AGPL logo

Comment (legalsylvain) [inline stock_picking_mass_action/README.rst:4]: @yvaucher : Done.

File: stock_picking_mass_action/wizard/mass_action.py [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+        return self.env.context.get('transfer', False)
+
+    @api.model
+    def _default_create_invoice(self):
+        return self.env.context.get('create_invoice', False)
+
+    confirm = fields.Boolean(
+        string='Mark as Todo', default=True,
+        help="""check this box if you want to mark as Todo the"""
+        """ the selected Pickings.""")
```
Comment (guewen) [inline stock_picking_mass_action/wizard/mass_action.py]: You don't need `"""` here, simple `"` work well:

``` python
    confirm = fields.Boolean(
        string='Mark as Todo', default=True,
        help="check this box if you want to mark as Todo the"
             " the selected Pickings.""")
```

BTW you can remove one `the` ;-)

Comment (guewen) [inline stock_picking_mass_action/wizard/mass_action.py]: Same remark for all the other fields below

File: stock_picking_mass_action/wizard/mass_action.py (hunk 11 above)
Comment (guewen) [inline stock_picking_mass_action/wizard/mass_action.py:88]: Visually an empty line is not like a commented line. Empty lines break the code in logical segments on which the eyes can focus. You are allowed not to like them, but that's something we can't argue about.

File: stock_picking_mass_action/wizard/mass_action.py [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+                transfert_wizard.do_detailed_transfer()
+
+        # Get all pickings ready to invoice and invoice them if asked
+        if self.create_invoice:
+            domain = [('invoice_state', '=', '2binvoiced'),
+                      ('id', 'in', picking_ids)]
+            to_invoice_picking_lst = picking_obj.search(
+                domain, order='min_date')
+            ctx = self.env.context.copy()
+            ctx['active_ids'] = [x.id for x in to_invoice_picking_lst]
```
Comment (guewen) [inline stock_picking_mass_action/wizard/mass_action.py]: `to_invoice_picking_lst.ids` is simpler
(I don't think it is necessary to add a `_lst` suffix, we can use `to_invoice_pickings` which speaks by itself)

File: stock_picking_mass_action/wizard/mass_action.py [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+            return {
+                'name': _('Stock Invoice Onshipping'),
+                'view_type': 'form',
+                'view_mode': 'form',
+                'res_model': 'stock.invoice.onshipping',
+                'type': 'ir.actions.act_window',
+                'target': 'new',
+                'context': self.env.context,
```
Comment (guewen) [inline stock_picking_mass_action/wizard/mass_action.py]: is `context['active_model']` correct at this point?

**Comment by guewen** [comment]
This branch needs a rebase

**Comment by legalsylvain** [comment]
@guewen : Done.

**Comment by legalsylvain** [comment]
@guewen : https://github.com/OCA/stock-logistics-workflow/pull/136#r39013231 treated. 
thanks for your review. CC @lepistone, @yvaucher 
(sorry for the delay ;-) )

**Comment by bealdav** [comment]
nice feature :+1:  thanks @guewen and @legalsylvain

**Comment by hparfr** [comment]
Thanks 👍

**Comment by legalsylvain** [comment]
Thanks for your reviews.
